### PR TITLE
common/bitutil: added data compression algorithm

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This PR adds the data compression algorithm originally presented in https://github.com/ethereum/go-ethereum/pull/3749 to package bitutil
It has been slightly modified to make it easier to read (better variable names) and work with any data size (the original one only worked with powers of two >=2)

### CODE REVIEW

File: common/bitutil/compress.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
+// Copyright 2013 The Go Authors. All rights reserved.
```
Comment (Arachnid) [inline common/bitutil/compress.go]: 2017!

Comment (Arachnid) [inline common/bitutil/compress.go]: Wait, copyright the Go authors?

Comment (karalabe) [inline common/bitutil/compress.go]: I'll change the copyright notice. It should be us, not upstream Go. Will add some tests and push a second commit on top that fixes this too :)

File: common/bitutil/compress.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+requires knowledge of the decompressed data length. Compression works as follows:
+
+if data only contains zeroes,
+  CompressBytes(data) == nil
+otherwise if len(data) <= 1,
+ CompressBytes(data) == data
+otherwise:
+ CompressBytes(data) == append(CompressBytes(nonZeroBits(data)), nonZeroBytes(data)...)
+where
+ nonZeroBits(data) is a bit vector with len(data) bits (MSB first):
```
Comment (Arachnid) [inline common/bitutil/compress.go]: You could probably phrase this as "nonZeroBits(data) is a bit vector with len(data) bits, where the nth bit is set iff the nth byte of data is nonzero" without loss of clarity.

**Comment by karalabe** [comment]
@zsfelfoldi @Arachnid 

I've written a fuzzer against the compression algo. The compress/decompress cycle worked correctly, however it turned out that the decompressor itself had quite a number of bugs if it was given invalid input, making it vulnerable to forgeries.

The following bugs were discovered and fixed:

 * If the target data size was set to `0`, the decompression entered an infinite recursion. I managed to trigger a variation that did a goroutine stack overflow with the `0x30` input and a 10 minute hang / loop with the `0x0020` input.
   * Although having a `0` target is debatable, we should not choke on it. I've added an extra clause into the decompression so that we return immediately and don't do any processing in this case.
 * If the target was `1` byte, the old code returned a slice of the input data. This is potentially bad because the output references the same backing array as the input.
   * I've changed the code so even in the case of a `1` byte target, there is no GC link between the input and output buffers.
 * Previously if the target was `1`, the decompressor always returned the first input byte and set the number of "used bytes" to `1`. However if the first byte of the input was a zero byte, then that cannot be a data byte, since zeroes are not stored.
   * In this case the decompressor must not "use" the first byte, rather has to assume that the target of `1` is actually represented by a `nil` slice prefix.
 * The original decompression code iterated from `0` to `len(decomp)-1` (via a range loop), and for each byte checked if it was set in the header bitset, and if so, it pulled a data byte from the input buffer. However the bitset in the header (`nonZeroBits`) may have had further bits set, but ones that we plain ignored because we didn't want to decode them. This may lead to ambiguous decoding deepening on the output buffer size.
   * I've changed the loop so it iterates over all the bits of the non zero bitset and ensures that all set bits have a corresponding data byte to place into the destination buffer.
   * The above fix also needs a second check to ensure that we *have* enough output space in the first place, and error out otherwise. Added that too.
 * The decompression when moving "data" bytes (marked as set via the non zero bitset) from the input to the output array didn't check if those bytes were zero. They **cannot be zero**, since zero bytes cannot be stored by the compression algo. Accepting zero bytes causes non-bijective mapping between uncompressed and compressed data (i.e. compressing such an decompressed output will produce a different compressed blob).
   * The solution is to explicitly check for zero bytes before accepting them into the `decomp` array, and if they are zero, then invalidate the entire compressed dataset.
 * The original decompressor used non-negative return values to signal the number of bytes processed from the input array. However this promotes bad API usage, hinting to developers that they can do streaming decompression. This is not the case, as any valid compressed data stream might be interpreted as having a `nil` stream in front.
   * There is no way around this decompression ambiguity unless stream decoding is forbidden. I've modified the code so that the `processed bytes` are returned only internally, but an outside wrapper call discards it, and enforces that exactly all the input bytes are used for a decompression. This ensures that a zero bytes can only be decompressed from a `nil` slice, and nothing else.
 * The previous decompressor has a single "error" return value of `-1` to signal that the input array is too short. Per the above bugfixes, we have 3 more possible errors: "extra bytes on input", "target data size exceeded" and "zero byte in input content".
   * I've modified the error handling so that all errors are explicitly returned as typed errors, and not hidden behind a cryptic `-1`.

@Arachnid @zsfelfoldi PTAL

**Comment by zsfelfoldi** [comment]
@karalabe thanks for the fixes. Originally I assumed that no invalid input will ever be decoded (although in the light client case it might be a more risky assumption so I was also thinking about adding proper checks). Also, I agree about the "streaming" restriction, if you take a look at https://github.com/zsfelfoldi/go-ethereum/blob/bloombits/core/bloombits/utils.go you can see that I originally wrapped these functions too. When I moved them to package bitutil I only moved the inner functions (as you suggested) but now I see that the inner decompression func should not be exposed.
Btw I used another "wrapping" too: I only returned the compressed data if it was shorter than the original, otherwise I returned a copy of the original. Also, I only decompressed any data if it was shorter than the target size. Now that we've brought back the wrapper for decompress, do you think we should add this extra check here too? Since we are always working with a known decompressed size it really makes no sense to ever store anything longer than the original so I would wrap them in package bloombits anyway. Maybe it would make more sense to add this check here.

**Comment by karalabe** [comment]
I've added that logic to in a working copy now, but it introduces some ambiguity. E.g. `0x00` can be decoded from `data: nil, len: 1` and from `data: 0x00, len: 1` too, both being valid. Trying to figure out a way to prevent cases like this, otherwise it makes testing quite annoying.

**Comment by karalabe** [comment]
@zsfelfoldi I've internalized the encoders/decoders and modified the tests so they directly test these, opposed to potential wrappers that discard certain data that encodes into longer segments. This is needed so that the fuzzer can actually test the code properly without "coverage" beind hidden by the length discard.

I've also added the compression and decompression wrappers that ensure longer encodings result in the input array being returned opposed to a longer encoded version. PTAL

File: common/bitutil/compress.go [hunk 3]
```diff
@@ -81,27 +93,38 @@ func CompressBytes(data []byte) []byte {
...
 
-// DecompressBytes decompresses data with a known target size. In addition to the
-// decompressed output, the function returns the length of compressed input data
-// corresponding to the output as the input slice may be longer.
+// DecompressBytes decompresses data with a known target size. If the input data
+// matches the size of the target, it means no compression was done in the first
+// place.
 func DecompressBytes(data []byte, target int) ([]byte, error) {
-	out, size, err := decompressBytes(data, target)
+	if len(data) == target {
```
Comment (zsfelfoldi) [inline common/bitutil/compress.go]: I would also check for len(data) > target here (which should always fail).

**Pull request merged**